cpp/src/arrow/filesystem/localfs_benchmark.cc

```diff
@@ -0,0 +1,163 @@
+      ASSERT_OK(MakeTrivialParquetFile(path.ToString()));
+    }
+
+    if (cur_nesting_level == nesting_depth_) {
+      return;
+    }
+
+    for (size_t i = 0; i < num_dirs_; ++i) {
+      std::string dirname = "dir_";
+      format(i, [&dirname](util::string_view formatted) {
+        dirname.append(formatted.data(), formatted.size());
+      });
+      ASSERT_OK_AND_ASSIGN(auto path, cur_root_dir.Join(std::move(dirname)));
+      InitializeDatasetStructure(cur_nesting_level + 1, std::move(path));
+    }
+  }
+
+  Status MakeTrivialParquetFile(const std::string& path) {
+    FieldVector fields{field("val", int64())};
+    auto batch = random::GenerateBatch(fields, 1 /*num_rows*/, 0);
+    ARROW_ASSIGN_OR_RAISE(auto table, Table::FromRecordBatches({batch}));
+
+    std::shared_ptr<io::OutputStream> sink;
+    ARROW_ASSIGN_OR_RAISE(sink, fs_->OpenOutputStream(path));
+
+    RETURN_NOT_OK(parquet::arrow::WriteTable(*table, arrow::default_memory_pool(), sink,
+                                             1 /*num_rows*/));
+
+    return Status::OK();
+  }
+
+ protected:
+  std::unique_ptr<TemporaryDir> tmp_dir_;
+  std::unique_ptr<LocalFileSystem> fs_;
+
+  const size_t nesting_depth_ = 2;
+  const size_t num_dirs_ = 10;
+  const size_t num_files_ = 10000;
+};
+
+/// Benchmark for `LocalFileSystem::GetFileInfoGenerator()` performance.
+///
+/// The test function is executed for each combination (cartesian product)
+/// of input arguments tuple (directory_readahead, file_info_batch_size)
+/// to test both internal parallelism and batching.
+BENCHMARK_DEFINE_F(LocalFSFixture, AsyncFileDiscovery)
+(benchmark::State& st) {
+  for (auto _ : st) {
+    // Instantiate LocalFileSystem with custom options for directory readahead
+    // and file info batch size.
+    auto options = LocalFileSystemOptions::Defaults();
+    options.directory_readahead = static_cast<int32_t>(st.range(0));
+    options.file_info_batch_size = static_cast<int32_t>(st.range(1));
+    auto test_fs = make_unique<LocalFileSystem>(options);
+    // Create recursive FileSelector pointing to the root of the temporary
+    // directory, which was set up by the fixture earlier.
+    FileSelector select;
+    select.base_dir = tmp_dir_->path().ToString();
+    select.recursive = true;
+    auto file_gen = test_fs->GetFileInfoGenerator(std::move(select));
+    size_t total_file_count = 0;
+    // Trigger fetching from the generator and count all received FileInfo:s.
+    auto visit_fut =
+        VisitAsyncGenerator(file_gen, [&total_file_count](const FileInfoVector& fv) {
+          total_file_count += fv.size();
+          return Status::OK();
+        });
+    ASSERT_FINISHES_OK(visit_fut);
+    st.SetItemsProcessed(total_file_count);
+  }
+}
+BENCHMARK_REGISTER_F(LocalFSFixture, AsyncFileDiscovery)
+    ->ArgNames({"directory_readahead", "file_info_batch_size"})
+    ->ArgsProduct({{1, 2, 4, 8, 16}, {1, 10, 100, 1000}})
```

Comment:
Can we perhaps cut down on the number of generated benchmarks? For example:
```suggestion
    ->ArgsProduct({{1, 4, 16}, {100, 1000}})
```

(`1` for `file_info_batch_size` seems so obviously pessimal that it needn't be tested, what do you think?)

Reply:
Makes sense.